src/wp-includes/rest-api.php

```diff
@@ -1776,3 +1776,119 @@ function rest_filter_response_by_context( $data, $schema, $context ) {
 
 	return $data;
 }
+
+
+/**
+ * Outputs a header link to a REST resource
+ *
+ * @since 5.5.0
+ *
+ * @param string       $url URL.
+ */
+function rest_head_link( $url ) {
+	if ( ! empty( $url ) ) {
+		printf( '<link rel="alternate" type="application/json" href="%s" />', esc_url( $url ) );
+	}
+}
+
+/**
+ * Return the rest route for a post.
+ *
+ * @since 5.5.0
+ *
+ * @param WP_Post       $post Post Object.
+ * @return string           Route.
+ */
+function rest_get_route_for_post( $post ) {
+	if ( ! $post instanceof WP_Post ) {
+		return '';
+	}
+
+	$post_type = get_post_type_object( $post->post_type );
+	if ( ! $post_type ) {
+		return '';
+	}
+
+	$route = '';
+	$controller = $post_type->get_rest_controller();
+
+	// The only two controllers that we can detect are the Attachments and Posts controllers.
+	if ( ! empty( $controller )  && in_array( $controller, array( 'WP_REST_Attachments_Controller', 'WP_REST_Posts_Controller' ), true ) ) {
```

Comment:
The `$controller` might be empty, but it would fail the `in_array` check in that case. I would remove the first part of the condition:

```suggestion
	if ( in_array( $controller, array( 'WP_REST_Attachments_Controller', 'WP_REST_Posts_Controller' ), true ) ) {
```

Reply:
To clarify, controller will be a `WP_REST_Controller` instance, not a class. So we'd need to pass `get_class` to the `in_array` call.

I think though we should short circuit immediately if there is no controller available and return an empty string. I don't think we should support someone being able to create ( via the filter ) a "rest" route for a post when that post doesn't have a rest controller.

Reply:
> To clarify, controller will be a `WP_REST_Controller` instance, not a class. So we'd need to pass `get_class` to the `in_array` call.
> 
> I think though we should short circuit immediately if there is no controller available and return an empty string. I don't think we should support someone being able to create ( via the filter ) a "rest" route for a post when that post doesn't have a rest controller.

The issue with this is that, without the get_controller_class, if you don't have it, class would resolve to false...and more things wouldn't work.

Reply:
Can you explain further? If the class isn't defined, the REST API won't work anyways.

Reply:
'rest_controller_class' - REST API Controller class name. Default is 'WP_REST_Terms_Controller'. So, without get_controller_class, if I looked for the class, and saw it was false, I would assume it was the Terms(or Post) controller. This now verifies that a show_in_rest = true, but a controller_class = false means use the default controller, which is not evident otherwise.

Reply:
Sorry, I'm still not understanding. Could you share a code example of what would go wrong? If you don't specify a `controller_class` then `WP_Post_Type::get_rest_controller()` will return an instance of the default class, `WP_REST_Posts_Controller` which is the correct behavior.

Reply:
That is exactly what I mean...thus why the function is needed and makes sense.

Reply:
Then what is the issue with returning if the controller isn't available?

Reply:
There isn't one anymore...the original code I wrote had the issue. This fixes it.

Reply:
I'm not seeing a `if ( ! $controller ) return;` in that code path though, https://github.com/WordPress/wordpress-develop/pull/333/files#diff-a32e81b26e79371567b4b1824bc2a427R1813.